### PR TITLE
Ensure queue isn't empty before trying to get a new module

### LINKED ISSUE #940: pool_size is throwing errors
## Expected Behavior 
No errors while using `pool_size`.

## Actual Behavior

```
mbaur@puppet-development:~ $ r10k deploy environment -p production
ERROR	 -> Thread error during concurrent module deploy: queue empty
WARN	 -> Puppet Forge module 'puppet-staging' has been deprecated, visit https://forge.puppet.com/puppet/staging for more information.
ERROR	 -> Thread error during concurrent module deploy: queue empty
ERROR	 -> Thread error during concurrent module deploy: queue empty
ERROR	 -> Thread error during concurrent module deploy: queue empty
ERROR	 -> Thread error during concurrent module deploy: queue empty
```

## Steps to Reproduce the Problem

```
mbaur@puppet-development:~ $ cat r10k.yaml
---
:cachedir: /home/mbaur/.puppetlabs/opt/puppet/cache/r10k
:sources:
  control:
    basedir: /home/mbaur/.puppetlabs/etc/code/environments
    remote: https://github.com/example42/psick

:deploy:
  purge_whitelist:
  - resource_types

:pool_size: 5
mbaur@puppet-development:~ $ rm -rf /home/mbaur/.puppetlabs/opt/puppet/cache/r10k/* /home/mbaur/.puppetlabs/etc/code/environments/*
mbaur@puppet-development:~ $ r10k deploy environment -p production
ERROR	 -> Thread error during concurrent module deploy: queue empty
WARN	 -> Puppet Forge module 'puppet-staging' has been deprecated, visit https://forge.puppet.com/puppet/staging for more information.
ERROR	 -> Thread error during concurrent module deploy: queue empty
ERROR	 -> Thread error during concurrent module deploy: queue empty
ERROR	 -> Thread error during concurrent module deploy: queue empty
ERROR	 -> Thread error during concurrent module deploy: queue empty
```

## Specifications

Please add this info:

  1. Output of ```facter -p``` on the failing node (at least the OS related facts)
```
mbaur@puppet-development:~ $ facter -p os
{"name"=>"Ubuntu", "family"=>"Debian", "release"=>{"major"=>"16.04", "full"=>"16.04"}, "lsb"=>{"distcodename"=>"xenial", "distid"=>"Ubuntu", "distdescription"=>"Ubuntu 16.04.6 LTS", "distrelease"=>"16.04", "majdistrelease"=>"16.04"}}
```
  1. Version of Puppet and of the module
```
mbaur@puppet-development:~ $ puppet --version
6.4.2
```
Module Version see --> https://github.com/example42/psick/blob/production/Puppetfile

  1. The relevant Puppet code and eventually Hiera data

See `Steps to Reproduce the Problem`

  1. Ruby version

```
mbaur@puppet-development:~ $ ruby --version
ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-linux]
```

Same for 2.5.3

### PR DESCRIPTION
**Pull request by baurmatt**
Fixes #940.

### CODE REVIEW

**Comment by baurmatt** [comment]
Hey @oesgalha and @reidmv. As you originally implemented this option (Thanks for this btw! :)), could you please take a look if this makes sense? Thanks! :)

File: lib/r10k/puppetfile.rb [hunk 1]
```diff
@@ -195,7 +195,7 @@ def visitor_thread(visitor, mods_queue)
       begin
         while mod = mods_queue.pop(true) do mod.accept(visitor) end
       rescue ThreadError => e
-        logger.error _("Thread error during concurrent module deploy: %{message}") % {message: e.message}
+        logger.debug _("Module thread %{id} exiting: %{message}") % {message: e.message, id: Thread.current.object_id}
```
Comment (Sharpie) [inline lib/r10k/puppetfile.rb:212]: After re-reading the docs a bit, it looks like `ThreadError` is raised by `.pop(true)` when the queue is empty. Logging this at DEBUG level seems reasonable. However, we should add an additional rescue clause at the end of the block to catch and log errors that occur during `mod.accept`:

```ruby
rescue => e
  logger.error _("Error during concurrent deploy of module %{mod_name}: %{message}") %
    {mode_name: mod.name,
     message: e.message}
  Thread.exit
end
```

**Review by Sharpie — CHANGES_REQUESTED** [review]

File: lib/r10k/puppetfile.rb (hunk 1 above)
Comment (reidmv) [reply lib/r10k/puppetfile.rb:212]: Added additional rescue block

**Review by Sharpie — APPROVED** [review]
Looks good 👍

**Comment by reidmv** [comment]
@Sharpie actually, upon further thought I'm not sure catching non-ThreadErrors here is a good idea... we should at least re-raise. Seems like there may be a danger of threads all terminating before finishing the queue of modules and (if we're catching like this here) r10k happily continuing without having processed all modules.

Seems like it'd be better to let the exception float up further.

Thoughts?

**Comment by oesgalha** [comment]
I think the behavior should be similar to the serial install.

If an error during `mod.accept` interrupts r10k in a serial install, then concurrent install should be interrupted too.

If this is the desired behavior, see [abort_on_exception](https://ruby-doc.org/core-2.6.3/Thread.html#method-i-abort_on_exception-3D) and this relevant doc:

> However, it's important to note that an exception that occurs in any thread except the main thread depends on abort_on_exception. This option is false by default, meaning that any unhandled exception will cause the thread to terminate silently when waited on by either join or value. You can change this default by either abort_on_exception= true or setting $DEBUG to true.

https://ruby-doc.org/core-2.6.3/Thread.html#class-Thread-label-Exception+handling

I think that re-raising with `Thread.main.raise(e)` should work too. An alternative to `abort_on_exception`.

**Comment by Sharpie** [comment]
Yeah, if we want to interrupt the main thread, then a simple re-raise won't cut it if we're generating and raising exceptions inside the worker threads:

```
$ cat test.rb
#!/bin/env ruby

Thread.new do
  raise 'boom!'
end

puts 'you should never see this'

raise 'lets try that again'

puts 'you definitely should never see this'

$ ruby test.rb
you should never see this
test.rb:9:in `<main>': lets try that again (RuntimeError)
```

**Comment by Sharpie** [comment]
Replacing `Thread.exit` with `Thread.main.raise(e)` the final `rescue => e` block seems to me like it would accomplish the goal of halting r10k.

One thing I'm not sure about is what would happen if the other threads are terminated abruptly while in the middle of pulling a `.git` update. I would hope that libgit2 is smart enough to handle this, but as they say: hope is not a strategy.

**Comment by Sharpie** [comment]
@reidmv The approach of storing the exception and waiting for the threads to finish looks great!

**Comment by baurmatt** [comment]
Hey @reidmv / @Sharpie, could you trigger a 3.3.1 release which includes this fix? Would love to use it in production! :)

**Comment by Magisus** [comment]
I've just released this as 3.3.1. Enjoy!

**Pull request merged**